### PR TITLE
Remove duplication in serving with and without graceful shutdown

### PR DESCRIPTION
**Pull request by jplatte**
Simpler implementation of the same idea from #2478, against latest `main`.

### CODE REVIEW

**Review by mladedav — APPROVED** [review]
This is great.

I just wonder if this couldn't go one step further and merge the `Serve` and `WithGracefulShutdown` into a single generic type. As it stands now, `Serve<M, S>` is just WithGracefulShutdown<M, S, Pending>` so why not embrace it?

The only difference I know of is that with the current separation users cannot call `with_graceful_shutdown` multiple times but I don't think that would be a problem.

**Comment by jplatte** [comment]
Sure, could do that. But it feels like leaking implementation details into the public API, and I don't see the little bit of extra code as much of a problem.

**Comment by mladedav** [comment]
I've taken a look.  It's because the graceful shutdown serve code has additional tracing with `trace!("connection {remote_addr} accepted");` and another one for closing the connection, but this test tries to parse each tracing line as a json with excpected fields.

So the easiest solutions are to pass the filter in `tracing_helpers.rs` as a function argument and pass `axum::rejection=trace` instead of the current `axum=trace` filter.

Right now the helper is used from just this one test so it could also just be changed in place, but passing the filter in is a bit cleaner.

**Comment by jplatte** [comment]
Thanks for investigating!
I've pushed a potentially overengineered fix, let me know what you think 😄

**Review by mladedav — APPROVED** [review]
Chaining the `with_filter` call certainly looks better on the callsite than just passing it in, so this is good with me. Even if a custom future impl maybe is a bit complex solution for a test helper.

File: axum/src/test_helpers/tracing_helpers.rs [hunk 1]
```diff
@@ -14,36 +21,69 @@ pub(crate) struct TracingEvent<T> {
...
+        f,
+        filter: None,
+        _phantom: PhantomData,
+    }
+}
+
+pub(crate) struct CaptureTracing<T, F> {
+    f: F,
+    filter: Option<Targets>,
+    _phantom: PhantomData<fn() -> T>,
```
[the remark this answers is not in the document]
Comment (jplatte) [reply axum/src/test_helpers/tracing_helpers.rs:38]: `fn(T)` would be contravariant, `fn() -> T` is covariant. It's just good form for generic "output" parameters, doesn't actually matter here. See https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns for details.

**Pull request merged**